YatimaStdLib/MultilinearPolynomial.lean

```diff
@@ -0,0 +1,162 @@
+    | some c => acc.insert b' (c + c')
+    | none => acc.insert b' c'
+
+instance : HAdd (MultilinearPolynomial α) (MultilinearPolynomial α)
+  (MultilinearPolynomial α) where hAdd x y := x.add y
+
+/--
+Multiplying two MLPs to obtain a MLP is not as straightforward because
+multiplication may increase the power of some variable if it's present on
+summands of the two initial MLPs, resulting on a polynomial that would no
+longer be linear.
+
+Thus we implement a "disjoint" multiplication, which considers that no variable
+is present on summands of both input MLPs.
+
+For example, `(x + 1) * (y + x + 2)` wouldn't be a disjoint multiplication, but
+`(x + 1) * (y + 2)` would.
+-/
+def disjointMul (mlp' : MultilinearPolynomial α) : MultilinearPolynomial α :=
+  mlp.foldl (init := default) fun pol b c =>
+    mlp'.foldl (init := pol) fun pol b' c' =>
+      pol.insert (b ||| b') (c * c')
+
+instance : HMul (MultilinearPolynomial α) (MultilinearPolynomial α)
+  (MultilinearPolynomial α) where hMul x y := x.disjointMul y
+
+/--
+Evaluates a MLP on an array whose values represent the values of the variables
+indexed from 0, matching the indexes of the array. Variables on indexes beyond
+the range of the array are considered to have value 0.
+-/
+def eval (input : Array α) : α :=
+  let inputSize := input.size
+  mlp.foldl (init := 0) fun acc b c => HAdd.hAdd acc $
+    Indices.ofBase b |>.foldl (init := c) fun acc i =>
+      acc * (if h : i < inputSize then input[i]'h else 0)
+
+/--
+Evaluates a MLP on a map that indicates the value of the variables indexed from
+0. Variables whose indexes aren't in the map are considered to have value 0.
+-/
+def eval' (input : Std.RBMap Nat α compare) : α :=
+  mlp.foldl (init := 0) fun acc b c => HAdd.hAdd acc $
+    Indices.ofBase b |>.foldl (init := c) fun acc i =>
+      acc * (input.find? i |>.getD 0)
+
+/-- Similar to `eval'`, but takes a list of `(index, value)` instead -/
+@[inline] def eval'L (input : List $ Nat × α) : α :=
+  mlp.eval' $ .ofList input _
+
+namespace Tests
+
+open LSpec
+
+-- TODO : prove this as a theorem
+#lspec check "roundtripping" $ ∀ n, (Indices.ofBase n).toBase = n
+
+/-- 3x₀x₄ + 2x₁ + 4 -/
+def pol1 := ofSummandsL [(2, [1]), (3, [4, 0]), (4, [])]
+
+/-- 2x₁x₄ + 4x₀x₄ + 1x₁x₃ + 3 -/
+def pol2 := ofSummandsL [(1, [1, 3]), (4, [4, 0]), (2, [4, 1]), (3, [])]
+
+/-- 9x₀x₄ + 6x₁ + 12 -/
+def pol1Scaled3 := ofSummandsL [(6, [1]), (9, [4, 0]), (12, [])]
+
+/-- 2x₁x₄ + 7x₀x₄ + 1x₁x₃ + 2x₁ + 7 -/
+def pol1AddPol2 :=
+  ofSummandsL [(1, [1, 3]), (2, [1]), (7, [4, 0]), (2, [4, 1]), (7, [])]
+
+/-- 4x₂x₃ + 12x₂ + 5 -/
+def pol3 := ofSummandsL [(12, [2]), (4, [2, 3]), (5, [])]
+
+/-- 12x₀x₂x₃x₄ + 36x₀x₂x₄ + 15x₀x₄ + 8x₁x₂x₃ + 16x₂x₃ + 24x₁x₂ + 48x₂ + 10x₁ + 20 -/
+def pol1MulPol3 := ofSummandsL [
+  (12, [0, 2, 3, 4]), (36, [0, 2, 4]), (15, [0, 4]),
+  (8, [1, 2, 3]), (24, [1, 2]), (10, [1]),
+  (16, [2, 3]), (48, [2]), (20, [])]
+
+#lspec
```

Comment:
Maybe add more tests to check for edge cases? Weird things come up all the time with basically anything having to do with the zero polynomial